// containers/pokemon/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import * as R from 'ramda'
import {Link} from 'react-router'

import {getPokemonById} from 'selectors'
import {fetchPokemonsById} from 'actions'

class Pokemon extends Component {
  componentDidMount () {
    this.props.fetchPokemonsById(this.props.params.id)
  }

  renderFields () {
    const {pokemon} = this.props
    const columnFields = R.compose(
      R.toPairs,
      R.pick([
        'type',
        'height',
        'weight',
        'candy',
        'egg'
      ])
    )(pokemon)

    return columnFields.map(([key, value]) => (
      <div className='column' key={key}>
        <div className='ab-details-title'>
          <p>{key}</p>
        </div>
        <div className='ab-details-info'>
          {value}
        </div>
      </div>
    ))
  }

  renderContent () {
    const {pokemon} = this.props

    return (
      <div className='thumbnail'>
        <div className='row'>
          <div className='col-md-6'>
            <img
              className='img-thumbnail'
              src={pokemon.image}
              alt={pokemon.name}
            />
          </div>
          <div className='col-md-6'>
            {this.renderFields()}
          </div>
        </div>
        <div className='caption-full'>
          <h4>{pokemon.name}</h4>
        </div>
      </div>
    )
  }

  renderSidebar () {
    const {pokemon} = this.props
    return (
      <div>
        <p className='lead'>Name</p>
        <div className='form-group'>
          <h1>{pokemon.name}</h1>
        </div>
        <Link to='/' className='btn btn-info btn-block'>Back to store</Link>
      </div>
    )
  }

  render () {
    const {pokemon} = this.props
    return (
      <div className='view-container'>
        <div className='container'>
          <div className='row'>
            <div className='col-md-9'>
              {pokemon && this.renderContent()}
            </div>
            <div className='col-md-3'>
              {pokemon && this.renderSidebar()}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    pokemon: getPokemonById(state, state.pokemonPage.id)
  }
}

const mapDispatchToProps = {
  fetchPokemonsById
}

export default connect(mapStateToProps, mapDispatchToProps)(Pokemon)
